src/java/datafu/pig/stats/entropy/Entropy.java

```
@@ -0,0 +1,350 @@
+ *
+ * input = LOAD 'input' AS (val: double);
+ *
+ * -- calculate the occurrence of each sample
+ * counts_g = GROUP input BY val;
+ * counts = FOREACh counts_g GENERATE COUNT(input) as cnt;
+ * 
+ * -- calculate entropy 
+ * input_counts_g = GROUP counts ALL;
+ * entropy = FOREACH input_counts_g GENERATE Entropy(counts) AS entropy; 
+ *
+ * </pre>
+ * </p>
+ */
+
+public class Entropy extends AccumulatorEvalFunc<Double> implements Algebraic {
+    
+    private static TupleFactory mTupleFactory = TupleFactory.getInstance();
+    
+    //entropy estimator for accumulator
+    //re-use the same entropy estimator for StreamingEntropy
+    private EntropyEstimator streamEstimator;
+    
+    public Entropy() {
+        //empirical estimator using Euler's number as logarithm base
+        streamEstimator = EntropyEstimator.createEstimator(EntropyEstimator.EMPIRICAL_ESTIMATOR, "");
+    }
+    
+    /*
+     * Algebraic implementation part
+     */
+    
+    @Override
+    public String getFinal() {
+        return Final.class.getName();
+    }
+
+    @Override
+    public String getInitial() {
+       return Initial.class.getName();
+    }
+
+    @Override
+    public String getIntermed() {
+        return Intermediate.class.getName();
+    }
+    
+    static public class Initial extends EvalFunc<Tuple> {
+        @Override
+        public Tuple exec(Tuple input) throws IOException {
+            Tuple t = mTupleFactory.newTuple(2);
+
+            try{
+                //input is a bag with one tuple containing
+                //the sample's occurrence frequency
+                DataBag bg = (DataBag) input.get(0);
+                Long cxl = null;
+                if(bg.iterator().hasNext()) {
+                   Tuple tp = bg.iterator().next();
+                   cxl = getFreq(tp);
+                }
+                
+                if(cxl == null || cxl.longValue() < 0) {
+                    //invalid input frequency
+                    t.set(0, null);
```

i wonder if we should throw an exception instead.  otherwise will people notice there is a problem?